In [1]:
import numpy as np
import random
import sys
sys.path.insert(0, '/Users/trimcao/Dropbox/Richardson/Fall-2017/cs6375-ml-ruozzi/solution/lib')
sys.path.insert(0, '/home/trimcao/Dropbox/Richardson/Fall-2017/cs6375-ml-ruozzi/solution/lib')
from DecisionTree import DecisionTree, BoostedTree

### Read in data

In [2]:
def read_in(file_path):
    X = []
    y = []
    with open(file_path, 'r') as f:
        for line in f:
            info = line.strip('\n').split(',')
            X.append([int(i) for i in info[1:]])
            y.append(int(info[0]))
    X = np.array(X)
    y = np.array(y)
    # make the labels -1 and 1
    y[y==0] = -1
    return X, y

In [3]:
X, y = read_in('hw3_data/heart_train.data')

In [4]:
X_test, y_test = read_in('hw3_data/heart_test.data')

In [5]:
N = X.shape[0]
weights = np.array([1/N for i in range(N)])

In [6]:
boost = BoostedTree()
boost.fit(X, y)

In [7]:
tree1 = boost.gen_tree1([10, 12, 14], X, y)
tree2 = boost.gen_tree2([10, 12, 14], X, y)
tree3 = boost.gen_tree3([10, 12, 14], X, y)
tree4 = boost.gen_tree4([10, 12, 14], X, y)
tree5 = boost.gen_tree5([10, 12, 14], X, y)

In [8]:
preds = tree1.predict(X)
print(np.mean(preds==y))

0.7875


In [9]:
preds = tree2.predict(X)
print(np.mean(preds==y))

0.7875


In [10]:
preds = tree3.predict(X)
print(np.mean(preds==y))

0.7875


In [11]:
preds = tree4.predict(X)
print(np.mean(preds==y))
preds = tree4.predict(X)
print(np.mean(preds==y))

0.625
0.625


### Train a decision tree with unlimited depth

In [8]:
clf = DecisionTree()
clf.fit(X, y, weights=weights)

In [9]:
print(clf.feature_dict)

{0: {0, 1}, 1: {0, 1}, 2: {0, 1}, 3: {0, 1}, 4: {0, 1}, 5: {0, 1}, 6: {0, 1}, 7: {0, 1}, 8: {0, 1}, 9: {0, 1}, 10: {0, 1}, 11: {0, 1}, 12: {0, 1}, 13: {0, 1}, 14: {0, 1}, 15: {0, 1}, 16: {0, 1}, 17: {0, 1}, 18: {0, 1}, 19: {0, 1}, 20: {0, 1}, 21: {0, 1}}


In [10]:
print(clf.predict_single(X_test[2]))

-1


In [11]:
preds = clf.predict(X_test)

In [12]:
print(np.mean(preds==y_test))

0.743315508021


In [13]:
print(clf.depth)

22


In [14]:
train_preds = clf.predict(X)
print(np.mean(train_preds==y))

0.9375


Depth = 22 means overfit.

### Train a decision tree with depth = 2

In [15]:
clf = DecisionTree(depth=2)
clf.fit(X, y, weights)

In [16]:
preds2 = clf.predict(X_test)

In [17]:
print(np.mean(preds2==y_test))

0.711229946524


Accuracy is not bad at all (0.711229946524).

### Probability with weights

In [18]:
weights = np.array([0.2, 0.5, 0.2, 0.1])
y_small = np.array([1, 0, 1, 0])
x_small = np.array([1, 1, 0, 0])

### adaBoost Decision Trees

In [19]:
ada = BoostedTree(M=16, tree_depth=2)
ada.fit(X, y)

In [20]:
print(ada.stages)

[0.65496069116765809, 0.18219499053584284, 0.2611841928246717, 0.26937481930643531, 0.11408306018447992, 0.19791387631833413, 0.37558941109257121, 0.23510026918532564, 0.27996464205510352, 0.26383917107807681, 0.22450393372127778, 0.16703096817594282, 0.14011286748562629, 0.080296003220651016, 0.14619933450718098, 0.067881328253971765]


In [21]:
preds = ada.predict(X_test)
print(preds)

[ 1.  1.  1.  1.  1.  1.  1. -1.  1.  1.  1.  1.  1. -1.  1.  1.  1.  1.
 -1.  1.  1.  1.  1.  1. -1.  1.  1.  1.  1.  1.  1.  1.  1.  1. -1.  1.
 -1.  1.  1.  1.  1.  1.  1. -1. -1.  1.  1.  1.  1.  1.  1.  1.  1.  1.
  1.  1.  1. -1.  1.  1.  1.  1.  1.  1. -1.  1.  1.  1.  1.  1.  1.  1.
  1. -1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  1. -1.  1.  1.  1.  1.  1.
  1.  1.  1.  1.  1.  1. -1.  1.  1.  1.  1.  1.  1.  1. -1.  1.  1. -1.
  1. -1. -1. -1.  1.  1. -1.  1. -1.  1. -1.  1. -1. -1. -1.  1.  1. -1.
  1.  1. -1.  1.  1.  1. -1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  1.
  1. -1. -1.  1.  1.  1.  1. -1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  1.
  1.  1.  1.  1.  1. -1. -1.  1.  1. -1. -1.  1. -1. -1. -1. -1. -1. -1.
 -1. -1. -1.  1.  1. -1. -1.]


In [22]:
print(np.mean(preds==y_test))

0.807486631016


In [23]:
train_preds = ada.predict(X)
print(np.mean(train_preds==y))

0.8375
